planner/cascades/transformation_rules.go

```diff
@@ -198,3 +199,41 @@ func (r *EnumeratePaths) OnTransform(old *memo.ExprIter) (newExprs []*memo.Group
 	}
 	return newExprs, true, false, nil
 }
+
+// PushSelDownSort pushes the Selection down to the child of Sort.
+type PushSelDownSort struct {
+}
+
+// GetPattern implements Transformation interface. The pattern of this rule
+// is `Selection -> Sort`.
+func (r *PushSelDownSort) GetPattern() *memo.Pattern {
+	return memo.BuildPattern(
+		memo.OperandSelection,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandSort, memo.EngineTiDBOnly),
+	)
+}
+
+// Match implements Transformation interface.
+func (r *PushSelDownSort) Match(expr *memo.ExprIter) bool {
+	return true
+}
+
+// OnTransform implements Transformation interface.
+// It will transform `sel->sort->x` to `sort->sel->x`.
+func (r *PushSelDownSort) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	sel := old.GetExpr().ExprNode.(*plannercore.LogicalSelection)
+	sort := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalSort)
+	childGroup := old.Children[0].GetExpr().Children[0]
+
+	newSort := plannercore.LogicalSort{ByItems: sort.ByItems}.Init(sort.SCtx(), sort.SelectBlockOffset())
+	newSel := plannercore.LogicalSelection{Conditions: sel.Conditions}.Init(sel.SCtx(), sel.SelectBlockOffset())
```

Comment:
We can reuse the `sel` instead of building a new plan node actually? because its stats would not change. We should set stats of `newSort` to stats of `sel` now also.

Reply:
The stats has not been prepared here?  It seems we can reuse both `sort` and `sel`?

Reply:
Oh yes.